File: src/frontend/financeiro/lib/app/data/model/fin_parcela_receber_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:financeiro/app/infra/infra_imports.dart';
import 'package:financeiro/app/data/model/model_imports.dart';
import 'package:intl/intl.dart';
import 'package:financeiro/app/data/domain/domain_imports.dart';

class FinParcelaReceberModel {
	int? id;
	int? idFinLancamentoReceber;
	int? idFinChequeRecebido;
	int? idFinStatusParcela;
	int? idFinTipoRecebimento;
	int? numeroParcela;
	DateTime? dataEmissao;
	DateTime? dataVencimento;
	DateTime? dataRecebimento;
	DateTime? descontoAte;
	double? valor;
	double? taxaJuro;
	double? taxaMulta;
	double? taxaDesconto;
	double? valorJuro;
	double? valorMulta;
	double? valorDesconto;
	String? emitiuBoleto;
	String? boletoNossoNumero;
	double? valorRecebido;
	String? historico;
	FinStatusParcelaModel? finStatusParcelaModel;
	FinTipoRecebimentoModel? finTipoRecebimentoModel;

	FinParcelaReceberModel({
		this.id,
		this.idFinLancamentoReceber,
		this.idFinChequeRecebido,
		this.idFinStatusParcela,
		this.idFinTipoRecebimento,
		this.numeroParcela,
		this.dataEmissao,
		this.dataVencimento,
		this.dataRecebimento,
		this.descontoAte,
		this.valor,
		this.taxaJuro,
		this.taxaMulta,
		this.taxaDesconto,
		this.valorJuro,
		this.valorMulta,
		this.valorDesconto,
		this.emitiuBoleto,
		this.boletoNossoNumero,
		this.valorRecebido,
		this.historico,
		this.finStatusParcelaModel,
		this.finTipoRecebimentoModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'numero_parcela',
		'data_emissao',
		'data_vencimento',
		'data_recebimento',
		'desconto_ate',
		'valor',
		'taxa_juro',
		'taxa_multa',
		'taxa_desconto',
		'valor_juro',
		'valor_multa',
		'valor_desconto',
		'emitiu_boleto',
		'boleto_nosso_numero',
		'valor_recebido',
		'historico',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Numero Parcela',
		'Data Emissao',
		'Data Vencimento',
		'Data Recebimento',
		'Desconto Ate',
		'Valor',
		'Taxa Juro',
		'Taxa Multa',
		'Taxa Desconto',
		'Valor Juro',
		'Valor Multa',
		'Valor Desconto',
		'Emitiu Boleto',
		'Boleto Nosso Numero',
		'Valor Recebido',
		'Historico',
	];

	FinParcelaReceberModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idFinLancamentoReceber = jsonData['idFinLancamentoReceber'];
		idFinChequeRecebido = jsonData['idFinChequeRecebido'];
		idFinStatusParcela = jsonData['idFinStatusParcela'];
		idFinTipoRecebimento = jsonData['idFinTipoRecebimento'];
		numeroParcela = jsonData['numeroParcela'];
		dataEmissao = jsonData['dataEmissao'] != null ? DateTime.tryParse(jsonData['dataEmissao']) : null;
		dataVencimento = jsonData['dataVencimento'] != null ? DateTime.tryParse(jsonData['dataVencimento']) : null;
		dataRecebimento = jsonData['dataRecebimento'] != null ? DateTime.tryParse(jsonData['dataRecebimento']) : null;
		descontoAte = jsonData['descontoAte'] != null ? DateTime.tryParse(jsonData['descontoAte']) : null;
		valor = jsonData['valor']?.toDouble();
		taxaJuro = jsonData['taxaJuro']?.toDouble();
		taxaMulta = jsonData['taxaMulta']?.toDouble();
		taxaDesconto = jsonData['taxaDesconto']?.toDouble();
		valorJuro = jsonData['valorJuro']?.toDouble();
		valorMulta = jsonData['valorMulta']?.toDouble();
		valorDesconto = jsonData['valorDesconto']?.toDouble();
		emitiuBoleto = FinParcelaReceberDomain.getEmitiuBoleto(jsonData['emitiuBoleto']);
		boletoNossoNumero = jsonData['boletoNossoNumero'];
		valorRecebido = jsonData['valorRecebido']?.toDouble();
		historico = jsonData['historico'];
		finStatusParcelaModel = jsonData['finStatusParcelaModel'] == null ? FinStatusParcelaModel() : FinStatusParcelaModel.fromJson(jsonData['finStatusParcelaModel']);
		finTipoRecebimentoModel = jsonData['finTipoRecebimentoModel'] == null ? FinTipoRecebimentoModel() : FinTipoRecebimentoModel.fromJson(jsonData['finTipoRecebimentoModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idFinLancamentoReceber'] = idFinLancamentoReceber != 0 ? idFinLancamentoReceber : null;
		jsonData['idFinChequeRecebido'] = idFinChequeRecebido != 0 ? idFinChequeRecebido : null;
		jsonData['idFinStatusParcela'] = idFinStatusParcela != 0 ? idFinStatusParcela : null;
		jsonData['idFinTipoRecebimento'] = idFinTipoRecebimento != 0 ? idFinTipoRecebimento : null;
		jsonData['numeroParcela'] = numeroParcela;
		jsonData['dataEmissao'] = dataEmissao != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataEmissao!) : null;
		jsonData['dataVencimento'] = dataVencimento != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataVencimento!) : null;
		jsonData['dataRecebimento'] = dataRecebimento != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataRecebimento!) : null;
		jsonData['descontoAte'] = descontoAte != null ? DateFormat('yyyy-MM-ddT00:00:00').format(descontoAte!) : null;
		jsonData['valor'] = valor;
		jsonData['taxaJuro'] = taxaJuro;
		jsonData['taxaMulta'] = taxaMulta;
		jsonData['taxaDesconto'] = taxaDesconto;
		jsonData['valorJuro'] = valorJuro;
		jsonData['valorMulta'] = valorMulta;
		jsonData['valorDesconto'] = valorDesconto;
		jsonData['emitiuBoleto'] = FinParcelaReceberDomain.setEmitiuBoleto(emitiuBoleto);
		jsonData['boletoNossoNumero'] = boletoNossoNumero;
		jsonData['valorRecebido'] = valorRecebido;
		jsonData['historico'] = historico;
		jsonData['finStatusParcelaModel'] = finStatusParcelaModel?.toJson;
		jsonData['finTipoRecebimentoModel'] = finTipoRecebimentoModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idFinLancamentoReceber = plutoRow.cells['idFinLancamentoReceber']?.value;
		idFinChequeRecebido = plutoRow.cells['idFinChequeRecebido']?.value;
		idFinStatusParcela = plutoRow.cells['idFinStatusParcela']?.value;
		idFinTipoRecebimento = plutoRow.cells['idFinTipoRecebimento']?.value;
		numeroParcela = plutoRow.cells['numeroParcela']?.value;
		dataEmissao = Util.stringToDate(plutoRow.cells['dataEmissao']?.value);
		dataVencimento = Util.stringToDate(plutoRow.cells['dataVencimento']?.value);
		dataRecebimento = Util.stringToDate(plutoRow.cells['dataRecebimento']?.value);
		descontoAte = Util.stringToDate(plutoRow.cells['descontoAte']?.value);
		valor = plutoRow.cells['valor']?.value?.toDouble();
		taxaJuro = plutoRow.cells['taxaJuro']?.value?.toDouble();
		taxaMulta = plutoRow.cells['taxaMulta']?.value?.toDouble();
		taxaDesconto = plutoRow.cells['taxaDesconto']?.value?.toDouble();
		valorJuro = plutoRow.cells['valorJuro']?.value?.toDouble();
		valorMulta = plutoRow.cells['valorMulta']?.value?.toDouble();
		valorDesconto = plutoRow.cells['valorDesconto']?.value?.toDouble();
		emitiuBoleto = plutoRow.cells['emitiuBoleto']?.value != '' ? plutoRow.cells['emitiuBoleto']?.value : 'S';
		boletoNossoNumero = plutoRow.cells['boletoNossoNumero']?.value;
		valorRecebido = plutoRow.cells['valorRecebido']?.value?.toDouble();
		historico = plutoRow.cells['historico']?.value;
		finStatusParcelaModel = FinStatusParcelaModel();
		finStatusParcelaModel?.descricao = plutoRow.cells['finStatusParcelaModel']?.value;
		finTipoRecebimentoModel = FinTipoRecebimentoModel();
		finTipoRecebimentoModel?.descricao = plutoRow.cells['finTipoRecebimentoModel']?.value;
	}	

	FinParcelaReceberModel clone() {
		return FinParcelaReceberModel(
			id: id,
			idFinLancamentoReceber: idFinLancamentoReceber,
			idFinChequeRecebido: idFinChequeRecebido,
			idFinStatusParcela: idFinStatusParcela,
			idFinTipoRecebimento: idFinTipoRecebimento,
			numeroParcela: numeroParcela,
			dataEmissao: dataEmissao,
			dataVencimento: dataVencimento,
			dataRecebimento: dataRecebimento,
			descontoAte: descontoAte,
			valor: valor,
			taxaJuro: taxaJuro,
			taxaMulta: taxaMulta,
			taxaDesconto: taxaDesconto,
			valorJuro: valorJuro,
			valorMulta: valorMulta,
			valorDesconto: valorDesconto,
			emitiuBoleto: emitiuBoleto,
			boletoNossoNumero: boletoNossoNumero,
			valorRecebido: valorRecebido,
			historico: historico,
		);			
	}

	
}

File: backend/financeiro/src/model/fin_parcela_receber_model.py
from src import db
from src.model.fin_status_parcela_model import FinStatusParcelaModel
from src.model.fin_tipo_recebimento_model import FinTipoRecebimentoModel


class FinParcelaReceberModel(db.Model):
    __tablename__ = 'fin_parcela_receber'

    id = db.Column(db.Integer, primary_key=True)
    numero_parcela = db.Column(db.Integer)
    data_emissao = db.Column(db.DateTime)
    data_vencimento = db.Column(db.DateTime)
    data_recebimento = db.Column(db.DateTime)
    desconto_ate = db.Column(db.DateTime)
    valor = db.Column(db.Float)
    taxa_juro = db.Column(db.Float)
    taxa_multa = db.Column(db.Float)
    taxa_desconto = db.Column(db.Float)
    valor_juro = db.Column(db.Float)
    valor_multa = db.Column(db.Float)
    valor_desconto = db.Column(db.Float)
    emitiu_boleto = db.Column(db.String(1))
    boleto_nosso_numero = db.Column(db.String(50))
    valor_recebido = db.Column(db.Float)
    historico = db.Column(db.Text)
    id_fin_lancamento_receber = db.Column(db.Integer, db.ForeignKey('fin_lancamento_receber.id'))
    id_fin_status_parcela = db.Column(db.Integer, db.ForeignKey('fin_status_parcela.id'))
    id_fin_tipo_recebimento = db.Column(db.Integer, db.ForeignKey('fin_tipo_recebimento.id'))

    fin_status_parcela_model = db.relationship('FinStatusParcelaModel', foreign_keys=[id_fin_status_parcela])
    fin_tipo_recebimento_model = db.relationship('FinTipoRecebimentoModel', foreign_keys=[id_fin_tipo_recebimento])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_fin_lancamento_receber = data.get('idFinLancamentoReceber')
        self.id_fin_cheque_recebido = data.get('idFinChequeRecebido')
        self.id_fin_status_parcela = data.get('idFinStatusParcela')
        self.id_fin_tipo_recebimento = data.get('idFinTipoRecebimento')
        self.numero_parcela = data.get('numeroParcela')
        self.data_emissao = data.get('dataEmissao')
        self.data_vencimento = data.get('dataVencimento')
        self.data_recebimento = data.get('dataRecebimento')
        self.desconto_ate = data.get('descontoAte')
        self.valor = data.get('valor')
        self.taxa_juro = data.get('taxaJuro')
        self.taxa_multa = data.get('taxaMulta')
        self.taxa_desconto = data.get('taxaDesconto')
        self.valor_juro = data.get('valorJuro')
        self.valor_multa = data.get('valorMulta')
        self.valor_desconto = data.get('valorDesconto')
        self.emitiu_boleto = data.get('emitiuBoleto')
        self.boleto_nosso_numero = data.get('boletoNossoNumero')
        self.valor_recebido = data.get('valorRecebido')
        self.historico = data.get('historico')

    def serialize(self):
        return {
            'id': self.id,
            'idFinLancamentoReceber': self.id_fin_lancamento_receber,
            'idFinChequeRecebido': self.id_fin_cheque_recebido,
            'idFinStatusParcela': self.id_fin_status_parcela,
            'idFinTipoRecebimento': self.id_fin_tipo_recebimento,
            'numeroParcela': self.numero_parcela,
            'dataEmissao': self.data_emissao.isoformat(),
            'dataVencimento': self.data_vencimento.isoformat(),
            'dataRecebimento': self.data_recebimento.isoformat(),
            'descontoAte': self.desconto_ate.isoformat(),
            'valor': self.valor,
            'taxaJuro': self.taxa_juro,
            'taxaMulta': self.taxa_multa,
            'taxaDesconto': self.taxa_desconto,
            'valorJuro': self.valor_juro,
            'valorMulta': self.valor_multa,
            'valorDesconto': self.valor_desconto,
            'emitiuBoleto': self.emitiu_boleto,
            'boletoNossoNumero': self.boleto_nosso_numero,
            'valorRecebido': self.valor_recebido,
            'historico': self.historico,
            'finStatusParcelaModel': self.fin_status_parcela_model.serialize() if self.fin_status_parcela_model else None,
            'finTipoRecebimentoModel': self.fin_tipo_recebimento_model.serialize() if self.fin_tipo_recebimento_model else None,
        }